PictureApp.API.Services/EmailNotificationService.cs

```diff
@@ -1,53 +1,44 @@
 using PictureApp.API.Providers;
+using PictureApp.Notifications;
 
 namespace PictureApp.API.Services
 {
     public class EmailNotificationService : INotificationService
     {
-        private readonly IRepository<NotificationTemplate> _repository;
         private readonly IEmailClientProvider _emailClientProvider;
+        private readonly INotificationTemplateService _notificationTemplateService;
 
-        public EmailNotificationService(IRepository<NotificationTemplate> repository,
-            IEmailClientProvider emailClientProvider)
+        public EmailNotificationService(IEmailClientProvider emailClientProvider, INotificationTemplateService notificationTemplateService)
         {
-            _repository = repository ?? throw new ArgumentNullException(nameof(repository));
             _emailClientProvider = emailClientProvider ?? throw new ArgumentNullException(nameof(emailClientProvider));
+            _notificationTemplateService = notificationTemplateService ?? throw new ArgumentNullException(nameof(notificationTemplateService));
         }
 
         public async Task SendAsync(string recipient, INotificationTemplateData templateData)
         {
-            // Attempt to get the template identified by an abbreviation
-            var notificationTemplate = _repository.Find(x => x.Abbreviation == templateData.TemplateAbbreviation)
-                .FirstOrDefault();
-            if (notificationTemplate == null)
+            if (templateData == null)
             {
-                throw new EntityNotFoundException(
-                    $"Can not find notification template with following abbreviation: {templateData.TemplateAbbreviation}");
+                throw new ArgumentNullException(nameof(templateData));
             }
-           
-            // Prepare the email body
-            var body = GetEmailBody(notificationTemplate.Body, templateData);
 
-            // Send the email via gateway client
-            await _emailClientProvider.SendAsync(new MailAddress(recipient), notificationTemplate.Subject, body);
-        }
+            MailAddress recipientMailAddress;
 
-        private string GetEmailBody(string template, INotificationTemplateData templateData)
-        {
-            var body = new StringBuilder(template);
+            try
+            {
+                recipientMailAddress = new MailAddress(recipient);
+            }
+            catch (FormatException e)
+            {
+                throw new ArgumentException($"Incorrect format for passed recipient email: `{recipient}`.", e);
+            }
 
-            templateData.GetKeys().ForEach(key => body.Replace(key, templateData.GetValue(key)));
-            return body.ToString();
+            var notification = await _notificationTemplateService.CreateNotification(templateData);
+
+            // Send the email via gateway client
+            await _emailClientProvider.SendAsync(recipientMailAddress, notification.Subject, notification.Body);
```

Comment:
one single question to solve my interest ;) send Async functionality is behind 'try and catch' block, its good way? What happens when sendAsync will crash?

Reply:
In this case 'try and catch' block is only for strict purpose - checking whether passed string represents recipient email is consistent with email format. The MailAddress class does not inlcude TryParse method, hence such approach has been taken. If exception appears on SendAsync than, the global error handler should handle it (I assume).